### PR TITLE
feat(desktop): rich tool rendering and permission modes (KAT-2090)

### PR DESCRIPTION
**Pull request by gannonh**
## Summary

Implements S02 for Kata Desktop with three focused commits:

- **KAT-2091**: Replaced generic tool cards with specialized rich renderers:
  - `FileEditCard` (Shiki diff rendering + diff stats + parse fallback)
  - `BashOutputCard` (ANSI terminal rendering + exit code status)
  - `FileReadCard` (syntax-highlighted file content + line metadata)
  - `WriteCard` (file preview with expandable full content)
  - `ToolCallCard` now dispatches by tool name and falls back to generic JSON
- **KAT-2092**: Extended `RpcEventAdapter` + shared types with structured tool args/results:
  - typed extraction for `edit`, `bash`, `read`, `write`
  - unknown tool passthrough without crashes
  - bash `tool_execution_update` support with `partialStdout`
  - expanded adapter tests with sample payloads and edge cases
- **KAT-2093**: Added permission modes + extension UI request/response bridge:
  - `permissionModeAtom` + `PermissionModeSelector` (`Explore` / `Ask` / `Auto`)
  - `PiAgentBridge` intercepts `extension_ui_request` and sends `extension_ui_response`
  - IPC + preload channels for extension UI and permission-mode updates
  - `ExtensionUIHandler` routes `confirm/select/input/notify`
  - `ToolApprovalDialog` with approve/reject and timeout auto-reject

## Validation

- `bun test apps/desktop/src/main/__tests__/rpc-event-adapter.test.ts`
- `bun test apps/desktop/src/main/__tests__/pi-agent-bridge.test.ts`
- `cd apps/desktop && bun run typecheck`
- `cd apps/desktop && bun run test`
- `cd apps/desktop && bun run build`

## Notes

- `pi-agent-bridge` tests intentionally exercise spawn-failure behavior; expected crash-path log output appears while tests still pass.
- Child tasks moved directly from Backlog to Done:
  - KAT-2091
  - KAT-2092
  - KAT-2093

Closes KAT-2090

<h3>Greptile Summary</h3>

This PR implements S02 for Kata Desktop across three focused sub-tasks: rich per-tool UI cards (KAT-2091), typed tool metadata extraction in the RPC adapter (KAT-2092), and a full extension UI request/response bridge with per-mode permission handling (KAT-2093). The changes are well-scoped and the key data paths are thoroughly tested.

**Highlights:**
- `ToolCallCard` now dispatches to specialized renderers (`FileEditCard`, `BashOutputCard`, `FileReadCard`, `WriteCard`) and falls back gracefully to generic JSON; `GenericToolCallCard` uses a "rendering failed" badge for all unrecognized tools, which is semantically misleading since no failure occurred — consider "generic" or "no renderer" instead
- `RpcEventAdapter` adds typed extraction with safe null-coalescing helpers; `countFromEdits` uses `filter(Boolean)` when tallying lines, silently dropping blank lines and understating the addition/deletion counts
- `ExtensionUIHandler` correctly handles the 'auto'/'explore'/'ask' modes using a `permissionModeRef` to avoid stale-closure bugs; the request-queue and timeout logic is sound
- `ToolApprovalDialog` uses an absolute deadline timestamp so interval restarts caused by prop reference churn are harmless
- Utility helpers (`asRecord`, `asString`, `asNumber`) are duplicated verbatim across all four new card components and already exist in `rpc-event-adapter.ts` — worth extracting to a shared module
- `PermissionModeSelector`'s `useEffect` fires on mount, sending an unnecessary `setPermissionMode('ask')` IPC call on every panel mount

<h3>Confidence Score: 5/5</h3>

Safe to merge; all remaining findings are P2 style/quality suggestions with no blocking correctness issues

No P0 or P1 issues found. The IPC lifecycle, stdin write guarding, stale-closure handling, and timeout countdown logic are all correct. The four findings are improvements to UX labelling, line-count accuracy, an unnecessary mount-time IPC call, and code duplication — none of which affect correctness or reliability of the feature.

No files require special attention; the "rendering failed" label in `ToolCallCard.tsx` and the `filter(Boolean)` in `rpc-event-adapter.ts` are the most visible items to address in a follow-up

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| apps/desktop/src/main/rpc-event-adapter.ts | Major expansion: adds typed extraction for edit/bash/read/write tools, language detection, and diff-line counting; `countFromEdits` uses `filter(Boolean)` which skips blank lines and slightly undercounts additions/deletions |
| apps/desktop/src/main/pi-agent-bridge.ts | Adds `extension_ui_request` interception via `dispatchRpcEvent`, `writeJsonLine` helper, `setPermissionMode`, and `sendExtensionUIResponse`; logic is sound with proper stdin writability checks |
| apps/desktop/src/main/ipc.ts | Wires up IPC handlers for extension UI request/response bridge and permission mode; clean cleanup in teardown and proper guard against destroyed renderer |
| apps/desktop/src/shared/types.ts | Adds well-structured typed interfaces for all tool args/results, ExtensionUIRequest subtypes, PermissionMode, and new IPC channels |
| apps/desktop/src/renderer/components/chat/ExtensionUIHandler.tsx | New component handling extension UI requests queue, per-mode auto-approve/reject, input/select modals, and toast notifications; `sendResponse`/`pushToast` in stale closure are safe due to functional state updaters |
| apps/desktop/src/renderer/components/chat/ToolApprovalDialog.tsx | Countdown timer dialog for confirm requests; interval uses absolute deadline so restarts from `onTimeout` prop changes are harmless, but a narrow race between user action and timeout expiry could send two responses for the same request |
| apps/desktop/src/renderer/components/chat/ToolCallCard.tsx | Now dispatches to specialized renderers; generic fallback shows "rendering failed" badge for any unrecognized tool, which is semantically misleading |
| apps/desktop/src/renderer/components/chat/FileEditCard.tsx | New component with three-tier diff rendering (explicit original/modified → parsed unified diff → reconstructed from edits array); parse fallback is well-thought-out |
| apps/desktop/src/renderer/components/chat/BashOutputCard.tsx | New component with ANSI terminal rendering, streaming partial stdout accumulation, and exit code display; utility helpers are duplicated from other card components |
| apps/desktop/src/renderer/components/chat/PermissionModeSelector.tsx | Three-mode toggle (Explore/Ask/Auto) synced to bridge via IPC; `useEffect` fires on mount sending a redundant `setPermissionMode('ask')` call |
| apps/desktop/src/renderer/atoms/chat.ts | Adds `partialStdout` accumulation on `tool_update` and clears it on `tool_end`; `ToolCallView` now uses typed `ToolArgs`/`ToolResult` |
| apps/desktop/src/main/__tests__/rpc-event-adapter.test.ts | Comprehensive new tests for edit/bash/read/write typed extraction, unknown tool passthrough, and streaming stdout; good coverage of edge cases |
| apps/desktop/src/preload/index.ts | Exposes three new IPC APIs (`onExtensionUIRequest`, `sendExtensionUIResponse`, `setPermissionMode`) through `contextBridge` with proper listener cleanup |

</details>

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Agent as Agent Subprocess
    participant Bridge as PiAgentBridge
    participant IPC as IPC Layer
    participant Renderer as ExtensionUIHandler
    participant Dialog as ToolApprovalDialog

    Note over Agent,Dialog: Normal tool execution flow
    Agent->>Bridge: tool_execution_start/update/end (JSON line)
    Bridge->>Bridge: dispatchRpcEvent()
    Bridge-->>IPC: emit('rpc-event', event)
    IPC->>Renderer: webContents.send(session:events)
    Renderer->>Renderer: applyChatEventAtom (accumulate partialStdout)

    Note over Agent,Dialog: Extension UI request flow
    Agent->>Bridge: extension_ui_request {id, method}
    Bridge->>Bridge: extractExtensionUIRequest()
    Bridge-->>IPC: emit('extension-ui-request', request)
    IPC->>Renderer: webContents.send(session:extension-ui-request)
    
    alt method == 'notify'
        Renderer->>Renderer: pushToast()
    else method == 'confirm' + mode == 'auto'
        Renderer->>IPC: sendExtensionUIResponse(id, {confirmed:true})
        IPC->>Bridge: ipcMain.handle(session:extension-ui-response)
        Bridge->>Agent: extension_ui_response (JSON line)
    else method == 'confirm' + mode == 'explore'
        Renderer->>IPC: sendExtensionUIResponse(id, {confirmed:false})
        IPC->>Bridge: ipcMain.handle
        Bridge->>Agent: extension_ui_response
    else method == 'confirm' + mode == 'ask'
        Renderer->>Dialog: open=true (queued confirm request)
        Dialog->>Dialog: countdown timer (5 min default)
        alt User approves
            Dialog->>Renderer: onApprove(id)
            Renderer->>IPC: sendExtensionUIResponse(id, {confirmed:true})
        else User rejects / timeout
            Dialog->>Renderer: onReject/onTimeout(id)
            Renderer->>IPC: sendExtensionUIResponse(id, {confirmed:false})
        end
        IPC->>Bridge: ipcMain.handle
        Bridge->>Agent: extension_ui_response
    else method == 'select' / 'input'
        Renderer->>Renderer: show modal (always, regardless of mode)
        Renderer->>IPC: sendExtensionUIResponse(id, {value})
        IPC->>Bridge: ipcMain.handle
        Bridge->>Agent: extension_ui_response
    end

    Note over Renderer,Dialog: Permission mode change
    Renderer->>IPC: ipcRenderer.invoke(session:permission-mode, mode)
    IPC->>Bridge: setPermissionMode(mode)
    Bridge->>Bridge: update in-memory permissionMode
```

<details><summary><h3>Comments Outside Diff (1)</h3></summary>

1. `apps/desktop/src/renderer/components/chat/PermissionModeSelector.tsx`, line 665-669 ([link](https://github.com/gannonh/kata/blob/a902188f2bc6635b3ddeee29c285e071a86b02bc/apps/desktop/src/renderer/components/chat/PermissionModeSelector.tsx#L665-L669)) 

   <a href="#">P2</a> **`setPermissionMode` fires redundantly on mount**

   The `useEffect` with `[mode]` as its dependency also fires on the initial render, sending a `setPermissionMode('ask')` IPC call even though the bridge defaults to `'ask'`. This is a harmless extra round-trip, but it means every time the `ChatPanel` mounts a spurious IPC call is made. Consider skipping the sync on initial mount:

   

   <details><summary>Prompt To Fix With AI</summary>

   `````markdown
   This is a comment left during a code review.
   Path: apps/desktop/src/renderer/components/chat/PermissionModeSelector.tsx
   Line: 665-669

   Comment:
   **`setPermissionMode` fires redundantly on mount**

   The `useEffect` with `[mode]` as its dependency also fires on the initial render, sending a `setPermissionMode('ask')` IPC call even though the bridge defaults to `'ask'`. This is a harmless extra round-trip, but it means every time the `ChatPanel` mounts a spurious IPC call is made. Consider skipping the sync on initial mount:

   

   How can I resolve this? If you propose a fix, please make it concise.
   `````
   </details>

</details>

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: apps/desktop/src/renderer/components/chat/ToolCallCard.tsx
Line: 37-41

Comment:
**"rendering failed" label applied to all unrecognized tools**

The `GenericToolCallCard` always renders a `rendering failed` badge, but no actual rendering failure has occurred — there's simply no specialized renderer registered for the tool. Future tools like `grep`, `glob`, `computer_use`, etc. would all display this badge while functioning correctly. Consider a label like `generic` or `no renderer` to distinguish the "fallback" case from an actual render error.

```suggestion
          <span className="rounded border border-slate-600 bg-slate-700/30 px-1.5 py-0.5 text-[10px] text-slate-400">
            generic
          </span>
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/desktop/src/main/rpc-event-adapter.ts
Line: 579-583

Comment:
**`filter(Boolean)` skips blank lines, undercounting additions/deletions**

`edit.newText.split('\n').filter(Boolean)` discards empty strings, so any blank lines in the edited content are silently excluded from the count. For example, adding a blank line at the end of a function would count as 0 additions. The same issue exists in `FileEditCard.tsx` in `buildFromEditsArray`.

```suggestion
      totals.additions += edit.newText.split('\n').length
      totals.deletions += edit.oldText.split('\n').length
```

If trailing newlines need to be excluded, consider using `.trimEnd().split('\n')` instead of `filter(Boolean)`.

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/desktop/src/renderer/components/chat/PermissionModeSelector.tsx
Line: 665-669

Comment:
**`setPermissionMode` fires redundantly on mount**

The `useEffect` with `[mode]` as its dependency also fires on the initial render, sending a `setPermissionMode('ask')` IPC call even though the bridge defaults to `'ask'`. This is a harmless extra round-trip, but it means every time the `ChatPanel` mounts a spurious IPC call is made. Consider skipping the sync on initial mount:

```suggestion
  const isMounted = useRef(false)

  useEffect(() => {
    if (!isMounted.current) {
      isMounted.current = true
      return
    }
    void window.api.setPermissionMode(mode).catch((error: unknown) => {
      console.error('[PermissionModeSelector] failed to update bridge permission mode', error)
    })
  }, [mode])
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: apps/desktop/src/renderer/components/chat/BashOutputCard.tsx
Line: 19-53

Comment:
**Utility helpers duplicated across all four card components**

`asRecord`, `asString`, and `asNumber` are defined identically (or near-identically) in `BashOutputCard.tsx`, `FileEditCard.tsx`, `FileReadCard.tsx`, and `WriteCard.tsx`, and the same helpers already exist in `rpc-event-adapter.ts`. Extracting them to a shared module (e.g., `@/lib/type-guards` or a co-located `card-utils.ts`) would eliminate ~40 lines of duplication and make future changes easier.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (1): Last reviewed commit: ["Merge remote-tracking branch &#39;origin/mai..."](https://github.com/gannonh/kata/commit/a902188f2bc6635b3ddeee29c285e071a86b02bc) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=26958318)</sub>

> Greptile also left **3 inline comments** on this PR.

## Release Notes

* **New Features**
  * Added permission mode selector to control how the agent handles interactive requests (explore, ask, or auto modes).
  * Added support for interactive extension UI dialogs, notifications, and input forms.
  * Enhanced tool output visualization with specialized displays for bash commands, file edits, reads, and writes.

* **UI Improvements**
  * Tool execution outputs now display with structured formatting, including command previews, file diffs, and line change summaries.

### CODE REVIEW

File: apps/desktop/src/renderer/components/chat/ExtensionUIHandler.tsx [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+          return
+        }
+
+        if (mode === 'explore') {
+          void sendResponse(request.id, { confirmed: false })
+          return
+        }
+      }
+
+      enqueue(request)
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/ExtensionUIHandler.tsx:187]: Addressed in `1bdcbb00`.

`ExtensionUIHandler` now auto-cancels unsupported `extension_ui_request` methods immediately (`{ cancelled: true }`) and returns early instead of enqueueing them. This prevents non-rendered methods from blocking the interactive queue.

File: apps/desktop/src/renderer/components/chat/ExtensionUIHandler.tsx [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+  }, [activeSelectRequest])
+
+  return (
+    <>
+      <ToolApprovalDialog
+        open={Boolean(activeConfirmRequest)}
+        request={activeConfirmRequest}
+        onApprove={(requestId) => {
+          void sendResponse(requestId, { confirmed: true })
+          setActiveRequest(null)
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/ExtensionUIHandler.tsx]: Fixed in `1bdcbb00`.

`sendResponse` now returns success/failure, and approve/reject/select/input handlers only clear `activeRequest` after a successful IPC send. On send failure the request stays open, so the user can retry and we avoid silently dropping the pending request.

File: apps/desktop/src/main/rpc-event-adapter.ts [hunk 3]
```diff
@@ -134,6 +182,190 @@ export class RpcEventAdapter {
...
+
+  private extractPartialStdout(event: RpcEvent): string | undefined {
+    const fromTopLevel = stringifyField(event.stdout)
+    if (fromTopLevel) {
+      return fromTopLevel
+    }
+
+    const resultRecord = asRecord(event.result)
+    const fromResult = stringifyField(resultRecord?.stdout)
+    if (fromResult) {
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/main/rpc-event-adapter.ts:364]: Updated in `1bdcbb00`.

`RpcEventAdapter.extractPartialStdout` now inspects streaming payloads from `partialResult` (top-level, `result.partialResult`, and `message.partialResult`) in addition to existing `stdout` fields, so bash updates emitted only through `partialResult` are surfaced to the renderer.

File: apps/desktop/src/renderer/components/chat/ToolCallCard.tsx [hunk 4]
```diff
@@ -30,7 +34,12 @@ export function ToolCallCard({ tool }: ToolCallCardProps) {
   return (
     <Collapsible.Root className="rounded-md border border-slate-700 bg-slate-900/60">
       <Collapsible.Trigger className="flex w-full items-center justify-between gap-3 px-3 py-2 text-left">
-        <span className="text-xs font-medium text-slate-100">{tool.name}</span>
+        <div className="flex min-w-0 items-center gap-2">
+          <span className="truncate text-xs font-medium text-slate-100">{tool.name}</span>
+          <span className="rounded border border-amber-500/40 bg-amber-500/15 px-1.5 py-0.5 text-[10px] text-amber-200">
+            rendering failed
+          </span>
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/ToolCallCard.tsx:41]: Good catch — changed in `1bdcbb00`.

For unknown tools the fallback card badge now says `generic` (neutral fallback) instead of `rendering failed`.

File: apps/desktop/src/main/rpc-event-adapter.ts [hunk 5]
```diff
@@ -267,3 +499,116 @@ export class RpcEventAdapter {
...
+
+  for (const line of lines) {
+    if (line.startsWith('+++ ') || line.startsWith('--- ') || line.startsWith('diff --git')) {
+      continue
+    }
+
+    if (line.startsWith('+')) {
+      additions += 1
+    } else if (line.startsWith('-')) {
+      deletions += 1
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/main/rpc-event-adapter.ts:605]: Addressed in `1bdcbb00`.

I replaced `split('\n').filter(Boolean)` counting with newline-aware `countTextLines` helpers so blank lines are preserved. This is now applied in both:
- `apps/desktop/src/main/rpc-event-adapter.ts` (`countFromEdits`)
- `apps/desktop/src/renderer/components/chat/FileEditCard.tsx` (`buildFromEditsArray` and explicit original/modified fallback counts)

Also added a regression test in `rpc-event-adapter.test.ts` covering blank-line counting for reconstructed edit results.

File: apps/desktop/src/renderer/components/chat/BashOutputCard.tsx [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+    try {
+      return JSON.stringify(value, null, 2)
+    } catch {
+      return String(value)
+    }
+  }
+
+  if (value === null || value === undefined) {
+    return ''
+  }
```
[the remark this answers is not in the document]
Comment (gannonh) [reply apps/desktop/src/renderer/components/chat/BashOutputCard.tsx]: Implemented in `1bdcbb00`.

I extracted shared card helpers into `apps/desktop/src/renderer/components/chat/toolCardUtils.ts` (`asRecord`, `asString`, `asNumber`, `asBoolean`, `countTextLines`) and updated `BashOutputCard`, `FileEditCard`, `FileReadCard`, and `WriteCard` to import from it.
